hilti/include/rt/types/set.h

```diff
@@ -25,33 +25,82 @@ namespace hilti::rt {
 namespace set {
 
 template<typename T>
-class SafeIterator
-    : public hilti::rt::detail::iterator::SafeIterator<Set<T>, typename Set<T>::SafeIterator, SafeIterator<T>> {
-public:
-    using Base = hilti::rt::detail::iterator::SafeIterator<Set<T>, typename Set<T>::SafeIterator, SafeIterator<T>>;
-    using Base::Base;
-};
+class Iterator {
+    using S = Set<T>;
+
+    std::weak_ptr<S*> _control;
+    typename S::V::iterator _iterator;
 
-template<typename T>
-class SafeConstIterator : public hilti::rt::detail::iterator::SafeIterator<const Set<T>, typename Set<T>::ConstIterator,
-                                                                           SafeConstIterator<T>> {
 public:
-    using Base =
-        hilti::rt::detail::iterator::SafeIterator<const Set<T>, typename Set<T>::ConstIterator, SafeConstIterator<T>>;
-    using Base::Base;
+    Iterator() = default;
+    Iterator(typename S::V::iterator iterator, const typename S::C& control)
+        : _control(control), _iterator(std::move(iterator)) {}
+
+    typename S::reference operator*() const {
+        if ( auto&& l = _control.lock() ) {
+            // Iterators to `end` cannot be dereferenced.
+            if ( _iterator == static_cast<const std::set<T>&>(**l).end() ) {
+                throw IndexError("iterator is invalid");
+            }
+
+            return *_iterator;
+        }
+
+        throw IndexError("iterator is invalid");
+    }
+
+    Iterator& operator++() {
+        if ( ! _control.lock() )
+            throw IndexError("iterator is invalid");
+
+        ++_iterator;
+        return *this;
+    };
+
+    Iterator operator++(int) {
+        auto ret = *this;
+        ++(*this); // Ensures the iterator is valid.
+        return ret;
+    }
+
+    friend bool operator==(const Iterator& a, const Iterator& b) {
+        if ( a._control.lock() != b._control.lock() )
+            throw InvalidArgument("cannot compare iterators into different sets");
+
+        return a._iterator == b._iterator;
+    }
+
+    friend bool operator!=(const Iterator& a, const Iterator& b) { return ! (a == b); }
 };
 
 } // namespace set
 
-/** HILTI's `Set` is an extended version `std::set`. */
+/** HILTI's `Set` is a `std::set`-like type with additional safety guarantees.
+ *
+ * In particular it guarantees that iterators are either valid, or throw an
```

Comment:
```suggestion
 * In particular it guarantees that iterators are either valid, or throws an
```

Reply:
Hmm, this needs to match _iterators_ so it should be plural.